src/azure-cli-core/azure/cli/core/msi_imds_authentication.py

```diff
@@ -0,0 +1,104 @@
+    use when authenticating to Azure AD. Mutually exclusive with client_id and msi_res_id.
+    - msi_res_id: Identifies, by ARM resource id, a specific explicit identity to use
+    when authenticating to Azure AD. Mutually exclusive with client_id and object_id.
+    - cloud_environment (msrestazure.azure_cloud.Cloud): A targeted cloud environment
+    - resource (str): Alternative authentication resource, default
+      is 'https://management.core.windows.net/'.
+    """
+
+    def __init__(self, **kwargs):
+        super(MSIImdsAuthentication, self).__init__(None)
+
+        self.identity_type, self.identity_id = None, None
+        temp = {k: v for k, v in kwargs.items() if k in ["client_id", "object_id", "msi_res_id"]}
+        if len(temp.keys()) > 1:
+            raise ValueError('"client_id", "object_id", "msi_res_id" are mutually exclusive')
+        elif len(temp.keys()) == 1:
+            self.identity_type, self.identity_id = next(iter(temp.items()))
+
+        self.cache = {}
+        self.resource = kwargs.get('resource')
+
+    def _set_token(self):
+        token_entry = self.get_token()
+        self.scheme, self.token = token_entry['token_type'], token_entry
+
+    def get_token(self):
+        import datetime
+        # let us hit the cache first
+        token_entry = self.cache.get(self.resource, None)
+        if token_entry:
+            expires_on = int(token_entry['expires_on'])
+            expires_on_datetime = datetime.datetime.fromtimestamp(expires_on)
+            expiration_margin = 5  # in minutes
+            if datetime.datetime.now() + datetime.timedelta(minutes=expiration_margin) <= expires_on_datetime:
+                logger.info("MSI: token is found in cache.")
+                return token_entry
+            logger.info("MSI: cache is found but expired within %s minutes, so getting a new one.", expiration_margin)
+            self.cache.pop(self.resource)
+
+        token_entry = self._retrieve_token_from_imds_with_retry()
+        self.cache[self.resource] = token_entry
+        return token_entry
+
+    def signed_session(self):
+        self._set_token()
+        return super(MSIImdsAuthentication, self).signed_session()
+
+    def _retrieve_token_from_imds_with_retry(self):
+        import random
+        import requests
+        import time
+
+        request_uri = 'http://169.254.169.254/metadata/identity/oauth2/token'
+        payload = {
+            'resource': self.resource,
+            'api-version': '2018-02-01'
+        }
+        if self.identity_id:
+            payload[self.identity_type] = self.identity_id
+
+        retry, max_retry = 1, 20
+        # simplified version of https://en.wikipedia.org/wiki/Exponential_backoff
+        slots = [100 * ((2 << x) - 1) / 1000 for x in range(max_retry)]
+        while retry <= max_retry:
+            result = requests.get(request_uri, params=payload, headers={'Metadata': 'true'})
```

Comment:
i will handle it

Reply:
Actually, after cross check, i don't think we need to to skip the cert verification, as the endpoint of imds is always HTTP, instead of HTTPS. I would bet, the `http://169.254.169.254` is a very special ip address which always get re-directed back to the host os in the same physical machine.